smart_card_connector_app/src/chrome-api-provider-unittest.js

```diff
@@ -0,0 +1,104 @@
+
+goog.scope(function() {
+
+const GSC = GoogleSmartCard;
+const ChromeApiProvider = GSC.ConnectorApp.ChromeApiProvider;
+
+const propertyReplacer = new goog.testing.PropertyReplacer();
+/** @type {!goog.testing.MockControl|undefined} */
+let mockControl;
+let mockServerMessageChannel;
+
+goog.exportSymbol('testChromeApiProvider', {
+  'setUp': function() {
+    mockControl = new goog.testing.MockControl();
+    mockServerMessageChannel =
+        new goog.testing.messaging.MockMessageChannel(mockControl);
+    for (const event
+             of ['onEstablishContextRequested',
+                 'onReleaseContextRequested',
+                 'onListReadersRequested',
+                 'onGetStatusChangeRequested',
+                 'onCancelRequested',
+                 'onConnectRequested',
+                 'onDisconnectRequested',
+                 'onTransmitRequested',
+                 'onControlRequested',
+                 'onGetAttribRequested',
+                 'onSetAttribRequested',
+                 'onStatusRequested',
+                 'onBeginTransactionRequested',
+                 'onEndTransactionRequested',
+    ]) {
+      // Mock chrome.smartCardProviderPrivate.*.addListener()/removeListener().
+      // Cast is to work around the issue that the return type of
+      // createFunctionMock() is too generic (as it depends on the value of the
+      // optional second argument).
+      const mockedAddListener = /** @type {!goog.testing.StrictMock} */ (
+          mockControl.createFunctionMock(`${event}.addListener`));
+      propertyReplacer.setPath(
+          `chrome.smartCardProviderPrivate.${event}.addListener`,
+          mockedAddListener);
+      const mockedRemoveListener = /** @type {!goog.testing.StrictMock} */ (
+          mockControl.createFunctionMock(`${event}.removeListener`));
+      propertyReplacer.setPath(
+          `chrome.smartCardProviderPrivate.${event}.removeListener`,
+          mockedRemoveListener);
+
+      // Check that a listener will be added to the event and the same listener
+      // will be removed.
+      chrome.smartCardProviderPrivate[event].addListener(
+          goog.testing.mockmatchers.isFunction);
+      let listener;
+      mockedAddListener.$once().$does(
+          (actual_listener) => {listener = actual_listener});
+      const argumentMatcher =
+          new goog.testing.mockmatchers.ArgumentMatcher(function(value) {
+            return value === listener;
+          }, 'verifySameListener');
+      chrome.smartCardProviderPrivate[event].removeListener(argumentMatcher);
+      mockedRemoveListener.$once();
+    }
+  },
+
+  'tearDown': function() {
+    // Check all mock expectations are satisfied.
+    mockControl.$verifyAll();
+  },
+
+  'testSmoke': function() {
+    mockControl.$replayAll();
+    const chromeApiProvider =
+        new ChromeApiProvider(mockServerMessageChannel, goog.Promise.resolve());
+    chromeApiProvider.dispose();
+  },
+});
+});  // goog.scope
```

Comment:
nit: Add a newline at the end of file, according to the usual text file conventions.

Reply:
done